components/script/dom/bluetoothpermissionresult.rs

```diff
@@ -0,0 +1,219 @@
+use dom::bindings::codegen::Bindings::NavigatorBinding::NavigatorBinding::NavigatorMethods;
+use dom::bindings::codegen::Bindings::PermissionStatusBinding::{PermissionState, PermissionStatusMethods};
+use dom::bindings::codegen::Bindings::WindowBinding::WindowBinding::WindowMethods;
+use dom::bindings::error::{Error, ErrorResult};
+use dom::bindings::inheritance::Castable;
+use dom::bindings::js::{JS, Root};
+use dom::bindings::reflector::{DomObject, reflect_dom_object};
+use dom::bindings::str::DOMString;
+use dom::bluetooth::{AsyncBluetoothListener, Bluetooth};
+use dom::bluetooth::{canonicalize_filter, get_allowed_devices, response_async, OPTIONS_ERROR};
+use dom::bluetoothdevice::BluetoothDevice;
+use dom::globalscope::GlobalScope;
+use dom::permissionstatus::{PermissionStatus, PermissionDescriptorType};
+use dom::promise::Promise;
+use ipc_channel::ipc::{self, IpcSender};
+use js::jsapi::JSContext;
+use std::rc::Rc;
+
+const DESCRIPTOR_TYPE_ERROR: &'static str = "Wrong type of descriptor in argument list.";
+
+// https://webbluetoothcg.github.io/web-bluetooth/#bluetoothpermissionresult
+#[dom_struct]
+pub struct BluetoothPermissionResult {
+    parent: PermissionStatus,
+    devices: DOMRefCell<Vec<JS<BluetoothDevice>>>,
+}
+
+impl BluetoothPermissionResult {
+    pub fn new_inherited() -> BluetoothPermissionResult {
+        BluetoothPermissionResult {
+            parent: PermissionStatus::new_inherited(),
+            devices: DOMRefCell::new(Vec::new()),
+        }
+    }
+
+    pub fn new(global: &GlobalScope) -> Root<BluetoothPermissionResult> {
+        reflect_dom_object(box BluetoothPermissionResult::new_inherited(),
+                           global,
+                           BluetoothPermissionResultBinding::Wrap)
+    }
+
+    fn get_bluetooth(&self) -> Root<Bluetooth> {
+        self.global().as_window().Navigator().Bluetooth()
+    }
+
+    fn get_bluetooth_thread(&self) -> IpcSender<BluetoothRequest> {
+        self.global().as_window().bluetooth_thread()
+    }
+
+    // https://w3c.github.io/permissions/#create-a-permissionstatus
+    pub fn create_from_descriptor(global: &GlobalScope,
+                                  descriptor: PermissionDescriptorType)
+                                  -> Root<BluetoothPermissionResult> {
+        let bt_permission_result = BluetoothPermissionResult::new(global);
+        bt_permission_result.upcast::<PermissionStatus>().set_query(descriptor);
+        bt_permission_result
+    }
+
+    #[allow(unrooted_must_root)]
+    // https://webbluetoothcg.github.io/web-bluetooth/#query-the-bluetooth-permission
+    pub fn permission_query(&self, promise: &Rc<Promise>, cx: *mut JSContext) {
+        let descriptor = self.parent.get_query().borrow();
+        let bluetoth_descriptor = match *descriptor {
+            PermissionDescriptorType::Bluetooth(ref d) => d,
+            _ => return promise.reject_error(cx, (Error::Type(DESCRIPTOR_TYPE_ERROR.to_owned()))),
```

Comment:
In the spec we already know, that this is a `BluetoothPermissionDescriptor`, because this should be a function parameter. The reason this is not passed as a parameter, that I got a borrow check error, when I tried to pass this argument. To avoid this I had to implement the Clone trait or a lot of generated struct, or reach this `BluetoothPermissionDescriptor ` from `self`, and i chose the second.